cf/data/data.py

```diff
@@ -3469,234 +3470,54 @@ def concatenate(cls, data, axis=0, _preserve=True):
-
-            for dim in data0.partitions.axes[1:]:
-
-                # Still here? Then see if there are any partition matrix
-                # boundaries to be created for this partition dimension
-                bounds0 = boundaries0[dim]
-                bounds1 = boundaries1[dim]
-
-                symmetric_diff = set(bounds0).symmetric_difference(bounds1)
-                if not symmetric_diff:
-                    # The partition boundaries for this partition
-                    # dimension are already the same in data0 and data1
-                    continue
-
-                # Still here? Then there are some partition boundaries to
-                # be created for this partition dimension in data0 and/or
-                # data1.
-                for f, g, bf, bg in (
-                    (data0, data1, bounds0, bounds1),
-                    (data1, data0, bounds1, bounds0),
-                ):
-                    extra_bounds = [i for i in bg if i in symmetric_diff]
-                    f.add_partitions(extra_bounds, dim)
-                # --- End: for
-            # --- End: for
-
-            # ------------------------------------------------------------
-            # 6. Concatenate data0 and data1 partition matrices
-            # ------------------------------------------------------------
-            #            if data0._flip != data1._flip:
-            if data0._flip() != data1._flip():
-                data0._move_flip_to_partitions()
-                data1._move_flip_to_partitions()
-
-            matrix0 = data0.partitions.matrix
-            matrix1 = data1.partitions.matrix
-
-            new_pmshape = list(matrix0.shape)
-            new_pmshape[0] += matrix1.shape[0]
-
-            # Initialise an empty partition matrix with the new shape
-            new_matrix = np.empty(new_pmshape, dtype=object)
-
-            # Insert the data0 partition matrix
-            new_matrix[: matrix0.shape[0]] = matrix0
-
-            # Insert the data1 partition matrix
-            new_matrix[matrix0.shape[0] :] = matrix1
-
-            data0.partitions.matrix = new_matrix
-
-            # Update the location map of the partition matrix of data0
-            data0.partitions.set_location_map((Paxis,), (axis,))
-
-            # ------------------------------------------------------------
-            # 7. Update the size, shape and dtype of data0
-            # ------------------------------------------------------------
-            #    original_shape0 = data0._shape
-
-            data0._size += data1._size
-
-            shape0 = list(shape0)
-            shape0[axis] += shape1[axis]
-            data0._shape = tuple(shape0)
-
-            dtype0 = data0.dtype
-            dtype1 = data1.dtype
-            if dtype0 != dtype1:
-                data0.dtype = np.result_type(dtype0, dtype1)
+        # Manage cyclicity of axes: if join axis was cyclic, it is no longer
+        cyclic_axes = data0._cyclic
```

Comment:
Hmm, that said, do we want to be concatenating along cyclic axes, anyway? As you know, the old code happily did so (without even updating the cyclicity).

However, I think your approach here is fine (consistent with 3.13.0, but better) - it'd be good to mention this behaviour in the docstring.

Reply:
> do we want to be concatenating along cyclic axes, anyway?

Good question. In light of considering it, may I suggest that we do allow it, with the behaviour as-is to unset any cyclicity on a cyclic join axis, but that we also *warn* (or "note*) about it via the logging, since users might then realise that joining via a cyclic axis isn't/wasn't what they wanted to or should do?

> However, I think your approach here is fine (consistent with 3.13.0, but better) - it'd be good to mention this behaviour in the docstring.

Indeed, regardless I will update the docstring to make it clear what the behaviour is, and how it has changed.

Reply:
And I will implement your suggestion from your first comment, which seems sensible in the context, with thanks for clarifying that. I assumed negative axes specifications worked based on one assertion in the test, but that may have worked by fluke...

Reply:
> do we want to be concatenating along cyclic axes, anyway?

Good question. In light of considering it, may I suggest that we do allow it, with the behaviour as-is to unset any cyclicity on a cyclic join axis, but that we also warn (or "note*) about it via the logging, since users might then realise that joining via a cyclic axis isn't/wasn't what they wanted to or should do?

> However, I think your approach here is fine (consistent with 3.13.0, but better) - it'd be good to mention this behaviour in the docstring.

Indeed, regardless I will update the docstring to make it clear what the behaviour is, and how it has changed.

> How about:
> ...

Looks good to me. I will implement your suggestion here, which seems sensible in the context, with thanks for clarifying that. I assumed negative axes specifications worked based on one assertion in the test, but that may have worked by fluke...

Reply:
Sounds good. 

I've realised why -ve axes worked before - `Data.cyclic` parses them. I guess I thought it didn't make sense to set  the cyclicity when it was already marked as non-cyclic (as per your original code), was thinking about the `axis in ...` code. It's six of one ...